### PR TITLE
Add I-JEPA

### PR DESCRIPTION
**Pull request by jmtzt**
# What does this PR do?

This PR adds [I-JEPA](https://github.com/facebookresearch/ijepa).

To-Do's:
- [x]  convert remaining checkpoints `ijepa_vith14_22k`, `ijepa_vith16_1k`, `ijepa_vitg16_22k`.
- [ ]  transfer checkpoints to the meta org

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @amyeroberts and @qubvel

File: src/transformers/models/ijepa/__init__.py
Comment (amyeroberts) [inline src/transformers/models/ijepa/__init__.py:60]: nit - formatting here looks a bit weird - our line length is 120 

```suggestion
    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure, module_spec=__spec__)
```

File: src/transformers/models/ijepa/configuration_ijepa.py
Comment (amyeroberts) [inline src/transformers/models/ijepa/configuration_ijepa.py:137]: onnx configs are now the responsibility of the optimum library 

```suggestion
```

File: tests/models/ijepa/test_modeling_ijepa.py
Comment (amyeroberts) [inline tests/models/ijepa/test_modeling_ijepa.py:205]: Has this been tested?

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (amyeroberts) [inline src/transformers/models/ijepa/modeling_ijepa.py:2]: copyright headers should be updated to reflect this model

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (amyeroberts) [inline src/transformers/models/ijepa/modeling_ijepa.py:690]: The input docstring here doesn't match the signature.

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model @jmtzt! 

Overall it's looking great. Did an initial review outlining some small things to update. 

Before merging, we'll need to make sure the slow model tests are passing. To trigger these, you'll need to push a commit (empty or otherwise) with the message `[run_slow] ijpea`. Me or another person at HF will then need to approve the workflow

**Comment by jmtzt** [comment]
Thanks for reviewing the code, @amyeroberts! :)

I've just pushed the adjustments, and now the fx support seems to be working fine.

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (amyeroberts) [inline src/transformers/models/ijepa/modeling_ijepa.py:112]: nit - standard in the library is to not have asserts in the modeling code 

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating - looks great! 

Just a small nit. It seems the slow model tests didn't run -- not really sure what happened there. Could you try with `[run-slow] ijepa` instead of `[run_slow] ijepa`? 

After than, just a merge conflict to resolve and we're good to go!

**Comment by jmtzt** [comment]
Thanks again for checking @amyeroberts :) I've just pushed the adjustments.

**Comment by amyeroberts** [comment]
Thanks for pushing again! Hmmmm - I'm not sure why the slow model tests aren't being picked up or run here cc @ydshieh

**Comment by jmtzt** [comment]
Thanks for looking into it. I fixed the merge conflicts, now the tests seem to have run without any issues. cc @amyeroberts

**Comment by amyeroberts** [comment]
@jmtzt Great! Final step is to move the checkpoints to be under the official organization and updating the paths in the PR. Could you share a link to all the checkpoints for this model under your name on hf.co?

**Comment by jmtzt** [comment]
Sure, here you can find them:

- [ijepa_vith14_1k](https://huggingface.co/jmtzt/ijepa_huge_patch14_1k)
- [ijepa_vith14_22k](https://huggingface.co/jmtzt/ijepa_huge_patch14_22k)
- [ijepa_vith16_1k](https://huggingface.co/jmtzt/ijepa_huge_patch16_1k)
- [ijepa_vitg16_22k](https://huggingface.co/jmtzt/ijepa_giant_patch16_22k)

**Comment by jmtzt** [comment]
@amyeroberts I've also updated the model cards :)

**Comment by qubvel** [comment]
Hi @jmtzt! Thanks for working on the model and sorry for the long wait, sometimes we are overwhelmed with issues and PRs! Don't hesitate to ping a few times!

**Comment by jmtzt** [comment]
No problem @qubvel, is there anything else I should do from my end? Thanks :)

**Review by qubvel — COMMENTED** [review]
Thanks for the contribution! The code looks good to me, here are some main questions I have

1. We are moving to a new conversion script pattern (see [mllama](https://github.com/huggingface/transformers/blob/04b4e441dcd0ed966e158c66dac870e67cab8ea8/src/transformers/models/mllama/convert_mllama_weights_to_hf.py)), can you please adapt your script too.
2. Modeling code is mostly reused from ViT model with with some minor changes. Despite it is not common for `transformers` to use inheritance, we recently introduced a new feature - modular transformers. With this new feature you actually can use inheritance, and then your code is going to be unpacked to a single model - single file format. More information and examples are in this [issue](https://github.com/huggingface/transformers/issues/33916). Would you have bandwidth to try the feature with this model? It should be a great candidate!
3. Regrading model cards, it would be great to add code snippets there!

Thanks, let me know if you need any further assistance or have any questions!

**Comment by NielsRogge** [comment]
Hi @jmtzt let us know if you need any help to push this one to the finish line!

**Comment by jmtzt** [comment]
> Hi @jmtzt let us know if you need any help to push this one to the finish line!

Hi @NielsRogge, I just need some time to go over the new conversion script and modelling pattern, I believe this weekend I can make some progress on it and keep you posted. Thanks!

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (jmtzt) [reply src/transformers/models/ijepa/modeling_ijepa.py:735]: that was an oversight from my side, since in the original `VisitonTransformerPredictor` from I-JEPA it's not used.

**Comment by jmtzt** [comment]
hey @qubvel just pushed the adjustments according to the new modular structure, as suggested. Will add the snippets to the model cards soon... :)

File: docs/source/en/model_doc/ijepa.md
Comment (qubvel) [inline docs/source/en/model_doc/ijepa.md:30]: It would be nice to have a code snippet here on how to make a prediction with the model! + adding the same snippet to model cards on the HF Hub

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:26]: should be modified in modular too, smth like

```python
_CHECKPOINT_FOR_DOC = "facebook/ijepa_vith14_1k"
```

**Review by qubvel — COMMENTED** [review]
Hi @jmtzt! Thanks for trying a new modular converter! Great job! I left a few comments below, but overall looks almost ready!

File: src/transformers/__init__.py
Comment (jmtzt) [reply src/transformers/__init__.py:1]: my local formatter overwrote the init file, I'll revert it

File: utils/check_docstrings.py
Comment (jmtzt) [reply utils/check_docstrings.py:334]: this is due to the modular model converter overwriting the `IJEPA_START_DOCSTRING` variable, leading to it not matching the function signature and when we do the repo consistency check, it fails, this happens similarly to the original ViT model, since it's also in this ignore list.

File: src/transformers/models/ijepa/modular_ijepa.py
Comment (jmtzt) [reply src/transformers/models/ijepa/modular_ijepa.py:155]: @qubvel I found out that changing this from `vit` to `ijepa` makes the `IJepaModelTest::test_model_outputs_equivalence` fail w/ the following error: `TypeError: 'tuple' object does not support item assignment`, not sure why tbh

File: docs/source/en/model_doc/ijepa.md
Comment (qubvel) [inline docs/source/en/model_doc/ijepa.md:54]: Do we have any checkpoints with pretrained classification heads?

File: src/transformers/models/ijepa/modular_ijepa.py
Comment (qubvel) [reply src/transformers/models/ijepa/modular_ijepa.py:155]: hmm, you could also change `base_model_prefix` accordingly, but they should be consistent to be able to load pretrained weights in both models

File: utils/check_docstrings.py
Comment (qubvel) [reply utils/check_docstrings.py:334]: I suppose with modular we can overwrite class/method docstring too, let me know if this is not the case

File: docs/source/en/model_doc/ijepa.md
Comment (jmtzt) [reply docs/source/en/model_doc/ijepa.md]: changed this to an image feature extraction example now...

**Comment by qubvel** [comment]
Thanks for iterating on the PR and addressing the comments!

While trying to use the converted checkpoint with the snippet from the model card I got the message that some weights were not initialized.

```python
import requests
from PIL import Image
from torch.nn.functional import cosine_similarity

from transformers import AutoModel, AutoProcessor

url_1 = "http://images.cocodataset.org/val2017/000000039769.jpg"
url_2 = "http://images.cocodataset.org/val2017/000000219578.jpg"
image_1 = Image.open(requests.get(url_1, stream=True).raw)
image_2 = Image.open(requests.get(url_2, stream=True).raw)

model_id = "jmtzt/ijepa_vith14_1k"
processor = AutoProcessor.from_pretrained(model_id)
model = AutoModel.from_pretrained(model_id)

def infer(image):
    inputs = processor(image, return_tensors="pt")
    outputs = model(**inputs)
    return outputs.pooler_output

embed_1 = infer(image_1)
embed_2 = infer(image_2)

similarity = cosine_similarity(embed_1, embed_2)
print(similarity)
```

Message:
Some weights of IJepaModel were not initialized from the model checkpoint at jmtzt/ijepa_vith14_1k and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']

Were these weights missed? Without the full model initialized I suppose the output similarity is incorrect, isn't it?

**Comment by jmtzt** [comment]
the original I-JEPA model doesn't have the pooling layer, so I think to get around this we might need to default the `add_pooling_layer` to `False` in its initialization, and modify this snippet accordingly to get the last hidden states rather than the `pooler_output`

**Comment by qubvel** [comment]
Ok, sounds good 👍

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:492]: Should be overwriten in `modular_`

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:662]: Also should be owerwritten in `modular_` because we don't have CLS token

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:712]: Probably we should also owerwrite the whole `forward` method in `modular_` for image classification model, because there is no CLS token I suppose it would be more correct to get average over tokens other than just first token. What do you think?

```python
logits = self.classifier(sequence_output.mean(dim=1))
```

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:646]: Should be overwritten in `modular_`

File: tests/models/ijepa/test_modeling_ijepa.py
Comment (qubvel) [inline tests/models/ijepa/test_modeling_ijepa.py:266]: Let's add `test_inference_interpolate_pos_encoding` to integration tests similar to what `Vit` model has.

File: docs/source/en/model_doc/ijepa.md
Comment (qubvel) [inline docs/source/en/model_doc/ijepa.md:52]:
```python
+ import torch

...
+ @torch.no_grad()
def infer(image):
    ...
```

**Review by qubvel — COMMENTED** [review]
Hi @jmtzt! Seems like almost everything is fine! Thanks for correcting snippets in docs and model cards! Snippets work fine on my side, I also did an experiment fine-tuning classification model - converges really good.

A few final nits regarding docstrings/constants and a suggestion regarding classification head and I will pass it to a core maintainers review, thank you for the great job!

**Comment by jmtzt** [comment]
hi @qubvel, thanks for the support and reviewing the PR! Just pushed the suggested changes, let me know if it's alright now.

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:680]: I suppose we can delete `_IMAGE_CLASS_CHECKPOINT` and `_IMAGE_CLASS_EXPECTED_OUTPUT` because we don't really have pretrained checkpoint for such a model

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (qubvel) [inline src/transformers/models/ijepa/modeling_ijepa.py:515]: hmm, it is not overwritten in modeling, while it exists in a modular file.. maybe a modular converter issue

**Review by qubvel — APPROVED** [review]
Thanks for addressing all the comments quickly! IMO it's ready for the next review

P.S. It may take a while since there are a lot of PRs for the final review in a line 🤗

File: src/transformers/models/ijepa/modeling_ijepa.py
Comment (jmtzt) [reply src/transformers/models/ijepa/modeling_ijepa.py:515]: in the `modular_model_converter.py` implementation, line 472, we have something like: 

 These top-level variables will always use the value in the `modular_xxx.py` file
`ASSIGNMENTS_TO_KEEP = {"_CHECKPOINT_FOR_DOC", }`

Comment (qubvel) [reply src/transformers/models/ijepa/modeling_ijepa.py:515]: ok, maybe we have to add "_EXPECTED_OUTPUT_SHAPE" too

File: tests/models/ijepa/test_modeling_ijepa.py
Comment (ArthurZucker) [reply tests/models/ijepa/test_modeling_ijepa.py:1]: Example of inheritance in tests with Gemma2 !

File: src/transformers/models/ijepa/modular_ijepa.py
Comment (ArthurZucker) [reply src/transformers/models/ijepa/modular_ijepa.py:43]: We should not even have to right it in the modular as it should be inherited!

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot for the PR! Modular makes it easy to understand: basically VIT but no cls token embedding right? (checked that no models in transformers already has this!)

File: src/transformers/models/ijepa/modular_ijepa.py
Comment (qubvel) [reply src/transformers/models/ijepa/modular_ijepa.py:43]: It was resolved because the method was updated. The above provided an already modified method, ViT has a CLS token, so the original method is slightly different.

e.g. ViT:
```
num_patches = embeddings.shape[1] - 1
num_positions = self.position_embeddings.shape[1] - 1
```
I-JEPA:
```
num_patches = embeddings.shape[1]
num_positions = self.position_embeddings.shape[1]
```

**Comment by ArthurZucker** [comment]
Feel free to ping me again for another review! 🫡

File: tests/models/ijepa/test_modeling_ijepa.py
Comment (jmtzt) [reply tests/models/ijepa/test_modeling_ijepa.py:1]: There are already integration tests implemented, quite similar to the ViT ones actually, the only one missing is the `test_inference_image_classification_head`, and IMO it doesn't make sense to have this, as we don't have a checkpoint for image classification tasks.

Comment (jmtzt) [reply tests/models/ijepa/test_modeling_ijepa.py:1]: also for your comment on the `interpolate_pos_encoding`, my implementation is different, to account for the lack of a CLS token...

**Comment by jmtzt** [comment]
> Feel free to ping me again for another review! 🫡

@ArthurZucker I've pushed the adjustments according to your comments, let me know if anything is missing. Thanks! :)

**Review by ArthurZucker — APPROVED** [review]
Very nice! Camel casing is a bit wrong but would look super ugly otherwise!

**Comment by qubvel** [comment]
@jmtzt congratulations on the model merged 🎉  was glad to collaborate with you on this!
Can you please confirm if we can transfer checkpoints to Facebook arg from your account?

**Comment by jmtzt** [comment]
thanks for the support @qubvel and @ArthurZucker :)

Sure, go ahead!

**Pull request merged**
